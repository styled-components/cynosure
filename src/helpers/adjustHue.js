// @flow

import parseColorString from '../internalHelpers/_parseColorString'
import rgbToHsl from '../internalHelpers/_rgbToHsl'
import hslToHex from '../internalHelpers/_hslToHex'
import hslToRgb from '../internalHelpers/_hslToRgb'

/**
 * Changes the hue of the color. Hue is a number between 0 to 360. The first
 * argument for adjustHue is the amount of degrees the color is rotated along
 * the color wheel.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   background: adjustHue('#448'),
 *   background: adjustHue('rgba(101,100,205,0.7)'),
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   background: ${adjustHue('#448')};
 *   background: ${adjustHue('rgba(101,100,205,0.7)')};
 * `
 *
 * // CSS in JS Output
 * element {
 *   background: "#5b4488";
 *   background: "rgba(136,100,205,0.7)";
 * }
 */
function adjustHue(degree: number, color: string): string {
  // parse color string to hsl
  const hslColor = rgbToHsl(parseColorString(color))
  const value = {
    ...hslColor,
    hue: (hslColor.hue + degree) % 360,
  }
  return value.alpha >= 1 || value.alpha === undefined
    ? hslToHex(value.hue, value.saturation, value.lightness)
    : `rgba(${hslToRgb(value.hue, value.saturation, value.lightness)},${value.alpha})`
}

export default adjustHue
