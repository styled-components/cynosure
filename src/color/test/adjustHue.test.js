// @flow
import adjustHue from '../adjustHue'

describe('adjustHue', () => {
  it('should adjustHue of a reduced hex color', () => {
    expect(adjustHue(20, '#448')).toEqual('#5b4488')
  })

  it('should adjustHue of a hex color', () => {
    expect(adjustHue(20, '#CCCD64')).toEqual('#a9cd64')
  })

  it('should adjustHue of an 8-digit hex color', () => {
    expect(adjustHue(20, '#6564CDB3')).toEqual('rgba(136,100,205,0.7)')
  })

  it('should adjustHue of an 4-digit hex color', () => {
    expect(adjustHue(20, '#0f08')).toEqual('rgba(0,255,85,0.53)')
  })

  it('should adjustHue of a color with opacity', () => {
    expect(adjustHue(20, 'rgba(101,100,205,0.7)')).toEqual(
      'rgba(136,100,205,0.7)',
    )
  })

  it('should adjustHue of a color and not go beyond 360', () => {
    expect(adjustHue(350, '#448')).toEqual('#444f88')
  })

  it('should adjustHue when passed a string for adjustment', () => {
    expect(adjustHue('20', '#448')).toEqual('#5b4488')
  })

  it('should return transparent when passed transparent', () => {
    expect(adjustHue('20', 'transparent')).toEqual('transparent')
  })
})
