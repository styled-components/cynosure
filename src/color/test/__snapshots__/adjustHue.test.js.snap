// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adjustHue should adjustHue of a color and not go beyond 360 1`] = `"#444f88"`;

exports[`adjustHue should adjustHue of a color with opacity 1`] = `"rgba(136,100,205,0.7)"`;

exports[`adjustHue should adjustHue of a hex color 1`] = `"#a9cd64"`;

exports[`adjustHue should adjustHue of a reduced hex color 1`] = `"#5b4488"`;

exports[`adjustHue should adjustHue when passed a string for adjustment 1`] = `"#5b4488"`;
